##  Converting Between Data Types

###  Learning Objectives
In this train, we will learn:

- How to convert from one data type to another using the **`CAST()`** function.



###  Overview
Every column in a database is assigned a **data type** when the database schema is created.  
When we query our data using the **`SELECT`** statement, the results are returned in the same data type as defined in the schema.

However, there are cases where we may want to **change the data type** of a column or an expression — for example, converting:

- A **string to a number**,  
- A **number to text**, or  
- A **date stored as text** into a proper **date format**.

The **`CAST()`** function allows us to perform such conversions directly in SQL.



###  Syntax
```sql
CAST(expression AS target_data_type)


In [1]:
%load_ext sql

##  Determining the Appropriate Precision

Before converting the data type of our `Est_population_in_millions` column,  
let’s first confirm the **data types** of all columns in our `Access_to_Basic_Services` table.

We can do this using the `SHOW COLUMNS` command.


In [4]:
%%sql
SHOW
    COLUMNS
FROM
    united_nations.Access_to_Basic_Services;


 * mysql+pymysql://root:***@localhost:3306/united_nations
10 rows affected.


[('Region', 'varchar(32)', 'YES', '', None, ''),
 ('Sub_region', 'varchar(25)', 'YES', '', None, ''),
 ('Country_name', 'varchar(37)', 'NO', '', None, ''),
 ('Time_period', 'int', 'NO', '', None, ''),
 ('Pct_managed_drinking_water_services', 'decimal(5,2)', 'YES', '', None, ''),
 ('Pct_managed_sanitation_services', 'decimal(5,2)', 'YES', '', None, ''),
 ('Est_population_in_millions', 'decimal(11,6)', 'YES', '', None, ''),
 ('Est_gdp_in_billions', 'decimal(8,2)', 'YES', '', None, ''),
 ('Land_area', 'decimal(10,2)', 'YES', '', None, ''),
 ('Pct_unemployment', 'decimal(5,2)', 'YES', '', None, '')]

From the output, we can see:
- `Country_name` → `VARCHAR`
- `Time_period` → `INT`
- `Est_population_in_millions` → `DECIMAL(11,6)`

This means that:
- The total **precision** is `11` (the total number of digits allowed),
- The **scale** is `6` (digits allowed *after* the decimal point).

This level of precision is accurate, but it makes the numbers too long and hard to read.


##  Finding the Maximum Value

To choose a better precision (the total number of digits),  
let’s first find the **maximum** population value in the column.


In [5]:
%%sql
SELECT
    MAX(Est_population_in_millions) AS Max_population
FROM
    united_nations.Access_to_Basic_Services;


 * mysql+pymysql://root:***@localhost:3306/united_nations
1 rows affected.


[(Decimal('1411.100000'),)]

Once we know the maximum value, we can decide on the total precision.

For example:
- If the maximum population is `1411.100000`,  
  then 4 digits before the decimal and 2 digits after it means we can use **DECIMAL(4,2)**.

So, our new format could be:
```sql
CAST(Est_population_in_millions AS DECIMAL(4,2))


##  Converting to the DECIMAL Data Type

Now that we’ve determined the most appropriate precision and scale  
(we will use `DECIMAL(6,2)` depending on the maximum population value),  
we can use the `CAST()` function to convert the `Est_population_in_millions` column.

This will round and format the values to make them easier to read,  
while still retaining sufficient accuracy for analysis.


In [6]:
%%sql
SELECT DISTINCT
    Country_name,
    Time_period,
    CAST(Est_population_in_millions AS DECIMAL(6,2)) AS Est_population_in_millions
FROM
    united_nations.Access_to_Basic_Services;


 * mysql+pymysql://root:***@localhost:3306/united_nations
1048 rows affected.


[('Kazakhstan', 2015, Decimal('17.54')),
 ('Kazakhstan', 2016, Decimal('17.79')),
 ('Kazakhstan', 2017, Decimal('18.04')),
 ('Kazakhstan', 2018, Decimal('18.28')),
 ('Kazakhstan', 2019, Decimal('18.51')),
 ('Kazakhstan', 2020, Decimal('18.76')),
 ('Kyrgyzstan', 2015, None),
 ('Kyrgyzstan', 2016, None),
 ('Kyrgyzstan', 2017, None),
 ('Kyrgyzstan', 2018, None),
 ('Kyrgyzstan', 2019, None),
 ('Kyrgyzstan', 2020, None),
 ('Tajikistan', 2015, Decimal('8.52')),
 ('Tajikistan', 2016, Decimal('8.73')),
 ('Tajikistan', 2017, Decimal('8.93')),
 ('Tajikistan', 2018, Decimal('9.13')),
 ('Tajikistan', 2019, Decimal('9.34')),
 ('Tajikistan', 2020, Decimal('9.54')),
 ('Turkmenistan', 2015, Decimal('5.77')),
 ('Turkmenistan', 2016, Decimal('5.87')),
 ('Turkmenistan', 2017, Decimal('5.97')),
 ('Turkmenistan', 2018, Decimal('6.07')),
 ('Turkmenistan', 2019, Decimal('6.16')),
 ('Turkmenistan', 2020, Decimal('6.25')),
 ('Uzbekistan', 2015, Decimal('31.30')),
 ('Uzbekistan', 2016, Decimal('31.85')),
 ('Uzbekistan', 2017, Decimal('32.39')),
 ('Uzbekistan', 2018, Decimal('32.96')),
 ('Uzbekistan', 2019, Decimal('33.58')),
 ('Uzbekistan', 2020, Decimal('34.23')),
 ('Afghanistan', 2015, Decimal('33.75')),
 ('Afghanistan', 2016, Decimal('34.64')),
 ('Afghanistan', 2017, Decimal('35.64')),
 ('Afghanistan', 2018, Decimal('36.69')),
 ('Afghanistan', 2019, Decimal('37.77')),
 ('Afghanistan', 2020, Decimal('38.97')),
 ('Bangladesh', 2015, Decimal('157.83')),
 ('Bangladesh', 2016, Decimal('159.78')),
 ('Bangladesh', 2017, Decimal('161.79')),
 ('Bangladesh', 2018, Decimal('163.68')),
 ('Bangladesh', 2019, Decimal('165.52')),
 ('Bangladesh', 2020, Decimal('167.42')),
 ('Bhutan', 2015, Decimal('0.74')),
 ('Bhutan', 2016, Decimal('0.75')),
 ('Bhutan', 2017, Decimal('0.76')),
 ('Bhutan', 2018, Decimal('0.76')),
 ('Bhutan', 2019, Decimal('0.77')),
 ('Bhutan', 2020, Decimal('0.77')),
 ('India', 2015, Decimal('1322.87')),
 ('India', 2016, Decimal('1338.64')),
 ('India', 2017, Decimal('1354.20')),
 ('India', 2018, Decimal('1369.00')),
 ('India', 2019, Decimal('1383.11')),
 ('India', 2020, Decimal('1396.39')),
 ('Iran (Islamic Republic of)', 2015, None),
 ('Iran (Islamic Republic of)', 2016, None),
 ('Iran (Islamic Republic of)', 2017, None),
 ('Iran (Islamic Republic of)', 2018, None),
 ('Iran (Islamic Republic of)', 2019, None),
 ('Iran (Islamic Republic of)', 2020, None),
 ('Maldives', 2015, Decimal('0.44')),
 ('Maldives', 2016, Decimal('0.45')),
 ('Maldives', 2017, Decimal('0.47')),
 ('Maldives', 2018, Decimal('0.49')),
 ('Maldives', 2019, Decimal('0.50')),
 ('Maldives', 2020, Decimal('0.51')),
 ('Nepal', 2015, Decimal('27.61')),
 ('Nepal', 2016, Decimal('27.86')),
 ('Nepal', 2017, Decimal('28.18')),
 ('Nepal', 2018, Decimal('28.51')),
 ('Nepal', 2019, Decimal('28.83')),
 ('Nepal', 2020, Decimal('29.35')),
 ('Pakistan', 2015, Decimal('210.97')),
 ('Pakistan', 2016, Decimal('213.52')),
 ('Pakistan', 2017, Decimal('216.38')),
 ('Pakistan', 2018, Decimal('219.73')),
 ('Pakistan', 2019, Decimal('223.29')),
 ('Pakistan', 2020, Decimal('227.20')),
 ('Sri Lanka', 2015, Decimal('20.97')),
 ('Sri Lanka', 2016, Decimal('21.20')),
 ('Sri Lanka', 2017, Decimal('21.44')),
 ('Sri Lanka', 2018, Decimal('21.67')),
 ('Sri Lanka', 2019, Decimal('21.80')),
 ('Sri Lanka', 2020, Decimal('21.92')),
 ('China', 2015, Decimal('1379.86')),
 ('China', 2016, Decimal('1387.79')),
 ('China', 2017, Decimal('1396.22')),
 ('China', 2018, Decimal('1402.76')),
 ('China', 2019, Decimal('1407.75')),
 ('China', 2020, Decimal('1411.10')),
 ("Democratic People's Republic of Korea", 2015, None),
 ("Democratic People's Republic of Korea", 2016, None),
 ("Democratic People's Republic of Korea", 2017, None),
 ("Democratic People's Republic of Korea", 2018, None),
 ("Democratic People's Republic of Korea", 2019, None),
 ("Democratic People's Republic of Korea", 2020, None),
 ('Japan', 2015, Decimal('127.14')),
 ('Japan', 2016, Decimal('127.08')),
 ('Japan', 2017, Decimal('126.97')),
 ('Japan', 2018, Decimal(

###  Explanation

- The `CAST()` function changes the data type of `Est_population_in_millions`  
  to a new **DECIMAL** with a total precision of `6` and a scale of `2`.
- This means each value can have up to **4 digits before** and **2 digits after** the decimal point.
- The resulting column is more readable and takes up less storage.

You can adjust the precision depending on the largest population value observed earlier.
For example:
- Use `DECIMAL(6,2)` for smaller values (under 1,000,000)
- Use `DECIMAL(9,2)` for larger values (hundreds of millions)


##  Summary

We have successfully performed the conversion, and our values now have only **two decimal places**,  
making the results more readable and efficient for analysis.

>  **Note:**  
> The `CAST()` function only changes the data type **within the result set** of the query.  
> It does **not** modify the actual data type stored in the database schema.

To permanently change a column’s data type in the database,  
you would need to use the `ALTER TABLE` statement — for example:

```sql
ALTER TABLE united_nations.Access_to_Basic_Services
ALTER COLUMN Est_population_in_millions DECIMAL(8,2);
